src/back/msl/writer.rs

```diff
@@ -668,7 +668,17 @@ impl<W: Write> Writer<W> {
                 write!(self.out, "{}", name)?;
             }
             crate::Expression::Load { pointer } => {
-                //write!(self.out, "*")?;
+                let should_dereference = match context.info[pointer].ty {
```

Comment:
We need to do something different here. In the IR, the `pointer` *has* to be a pointer already, so this needs to always dereference, unconditionally. However, I imagine that if we see that the expression is `Access` or `AccessIndex`, then we know that there was already a root of this access chain somewhere, and we inserted `->` in the beginning (#699), so we don't need to have a dereference here.

Reply:
Oh right, I see! Yeah I thought that dereferencing all the time was wrong because it breaks `main2Output`:
```metal

struct main2Output {
    metal::float4 gl_Position1 [[position]];
    float gl_PointSize1 [[point_size]];
    type5 gl_ClipDistance1 [[clip_distance]];
};
vertex main2Output main2(
) {
    gl_PerVertex _ = {};
    main1(_);
    const auto _tmp = type9 {*_.gl_Position, *_.gl_PointSize, {*_.gl_ClipDistance[0]}};
    return main2Output { _tmp.gl_Position1, _tmp.gl_PointSize1, {_tmp.gl_ClipDistance1[0]} };
}
```
but that just means we need more changes.

Reply:
Instead we want something like:
```metal
struct main2Output {
    metal::float4 gl_Position1 [[position]];
    float gl_PointSize1 [[point_size]];
    type3 gl_ClipDistance1 [[clip_distance]];
};
vertex main2Output main2(
) {
    gl_PerVertex _owned = {};
    auto _ = &_owned;
    main1(*_);
    const auto _tmp = type7 {(*_).gl_Position, (*_).gl_PointSize, {(*_).gl_ClipDistance[0]}};
    return main2Output { _tmp.gl_Position1, _tmp.gl_PointSize1, {_tmp.gl_ClipDistance1[0]} };
}
```

Reply:
Or actually just
```metal
vertex main2Output main2(
) {
    gl_PerVertex _owned = {};
    thread gl_PerVertex& _ = _owned;
    main1(_);
    const auto _tmp = type7 {_.gl_Position, _.gl_PointSize, {_.gl_ClipDistance[0]}};
    return main2Output { _tmp.gl_Position1, _tmp.gl_PointSize1, {_tmp.gl_ClipDistance1[0]} };
}
```

Reply:
Oh no, nevermind, simply
```metal
ertex main2Output main2(
) {
    gl_PerVertex _ = {};
    main1(_);
    const auto _tmp = type7 {_.gl_Position, _.gl_PointSize, {_.gl_ClipDistance[0]}};
    return main2Output { _tmp.gl_Position1, _tmp.gl_PointSize1, {_tmp.gl_ClipDistance1[0]} };
}
```
works.

Reply:
It shouldn't be dereferenced in the case of `LocalVariable` and `GlobalVariable` either (unless the IR is wrong) because otherwise in the collatz msl we get:
```metal
metal::uint collatz_iterations(
    metal::uint n_base
) {
    metal::uint n;
    metal::uint i = 0u;
    n = n_base;
    while(true) {
        if (*n <= 1u) {
            break;
        }
        if ((*n % 2u) == 0u) {
            n = *n / 2u;
        } else {
            n = (3u * *n) + 1u;
        }
        i = *i + 1u;
    }
    return *i;
}
```
which is pretty clearly wrong.

Reply:
After #693 is merged, `FunctionArgument`s are all passed by reference, and shouldn't be dereferenced either.

Reply:
We might not need this after the changes to #693.